message_parser/src/main/java/cumulus_message_adapter/message_parser/MessageParser.java

```diff
@@ -0,0 +1,68 @@
+package cumulus_message_adapter.message_parser;
+
+import com.amazonaws.services.lambda.runtime.Context; 
+
+
+/**
+ * Hanldes messages by passing the input through the message adapter and using the output as input 
+ * for business logic. The output of the business logic is passed back to the message adapter.
+ */
+public class MessageParser implements IMessageParser
+{
+    /**
+     * Instance of the message adapter to use for message processing
+     */
+    private IMessageAdapter _messageAdapter;
+
+    /**
+     * Default constructor - creates a new message adapter instance
+     */
+    public MessageParser()
+    {
+        _messageAdapter = new MessageAdapter();
+    }
+
+    /**
+     * Constructor, takes an instance of the message adapter. Can be used for testing
+     * 
+     * @param messageAdapter - instance of the message adapter to use
+     */
+    public MessageParser(IMessageAdapter messageAdapter)
+    {
+        _messageAdapter = messageAdapter;
+    }
+
+    /**
+     * Handles the message by passing the input through the message adapter with the context serialized as Json.
+     * Calls task business logic and passes the output back to the message adapter
+     * 
+     * @param input - input Json
+     * @param context - AWS Lambda context
+     * @param task - callback to business logic function
+     */
+    public String HandleMessage(String input, Context context, ITask task)
+    {
+        Boolean messageAdapterDisabled = Boolean.valueOf(System.getenv("CUMULUS_MESSAGE_ADAPTER_DISABLED"));
+
+        try
+        {
+            // If the message adapter is disabled, call the task with original input
+            if(messageAdapterDisabled)
+            {
+                return task.PerformFunction(input);
+            }
+
+            String remoteEvent = _messageAdapter.LoadRemoteEvent(input);
+
+            String eventInput = _messageAdapter.LoadNestedEvent(remoteEvent, context);
+
+            String taskOutput = task.PerformFunction(eventInput);
+
+            return _messageAdapter.CreateNextEvent(remoteEvent, eventInput, taskOutput);
+        }
+        catch(Exception e)
+        {
+            return "{\"payload\":null,\"exception\":\"" + e.getMessage() + "\"}";  
```

Comment:
If the exception is a descendant of "WorkflowError", it should set "payload=null" and "exception=messagetype" in the full remoteEvent and return that remote event.  (This is not returning the full event.)

If the error is of any other type, it should be raised instead of being caught here.

Reply:
See https://github.com/cumulus-nasa/cumulus-message-adapter-js/blob/master/index.js#L120 for a reference.  I copied that almost exactly from Patrick's original sled code.

Reply:
I can update it, but think it's kind of hacky. We don't have any workflow exceptions defined in Java so I'd have to catch all exceptions and then re-throw if I found WorkflowException in the string. I'll put it in, it just doesn't really go with the language.

Reply:
I put it in. I used WorkflowError and WorkflowException (which would be more Java-ish). I don't have an actual real-life use case for this right now, but plan on talking to podaac. Other exceptions will be re-thrown as a MessageAdapterException.